### DISCUSSION TITLE
Extended Analysis

### DISCUSSION THREAD

**Discussion by DaveSkender** [DaveSkender/Stock.Indicators#320] [Ideas]
For those of you who've been following the development of this library, before the initial v1.0, we had a few indicators that supplied additional analysis in the results; things like "IsBullish" and "IsReversal".  Based on feedback from users (#99), we decided to keep the library in a more basic form and excluded additional analysis values.

I'm open to adding these back since they are useful for identifying trade signals and conditions; however, I'd like to do it in a way that does not degrade performance or add complexity to the basic indicators.  We'd need to satisfy both groups of users 1) those who want basic, pure, and fast indicators, and 2) those who are willing to sacrifice a bit of performance for having extended information.

My initial thought was to have an entirely separate "Stock.Indicators.Extended" library, but there are probably several ways this could be implemented.

Thoughts?

**Comment by DaveSkender** [comment]
I think for cases where we add extended analysis, we can add `**Extended` variant methods in the current library.  This is a bit cleaner than adding a flag in the base indicator and simpler than having an entirely separate library.  This approach allows both the base indicator to be as simple and performant as possible and the extended indicator to include additional information.  For example, for [SMA](https://daveskender.github.io/Stock.Indicators/indicators/Sma/#content), the extended version would look like:

![image](https://user-images.githubusercontent.com/8432125/106371957-d9e64f80-6338-11eb-9cdf-7e4591c09600.png)

The base indicator would be:
```
IEnumerable<SmaResult> results = Indicator.GetSma(history, lookbackPeriod);
```
And would include a simpler result class only containing `Date` and `Sma` properties.

**Reply by DaveSkender** [reply]
I think I like the `**Analysis` suffix convention better (e.g. `GetSmaAnalysis`)

**Reply by jimtollan** [reply]
Dave - I responded to one of the other threads in support of this idea, but just wanted to further amplify that support here also. An *Extended* library, where it doesn't compromise the core Indicators library would be an absolute killer. This would literally allow myself and a multitude of users to backtest and explore new strategies in very consistent and organised fashion. 

Would love to know if there's a liklihood of this happening in the near future.

cheers
jim

**Reply by DaveSkender** [reply]
I don’t really have a timeline for the Roadmap #218 given I can really only work on this with unpredictable free time, but can at least let you see my priorities.  With that said, I’m always looking for help, to make this go a bit faster.  Contributors are welcome!

**Reply by jimtollan** [reply]
cheers for the reply. my technical abilities would preclude me from contributing too much, however, i would be more thn happy to test out any new functionality that was required.

**Comment by DaveSkender** [comment]
I think another API option here could look more like:

```csharp
# two steps
var results = quotes.GetSma(5);
var analysis = results.Analyze();

# also supports chaining
var analysis = quotes
    .GetSma(5)
    .Analyze();
```
This is a bit easier to use IMO and would keep base indicator calculations as separate concerns.  Thoughts and feedback?

**Reply by jimtollan** [reply]
it would be nice also (if it doesn't already exist) to have a further extension method (or methods) linked to IEnumerable<Quote>. one example that I can think straight off the bat that would be very useful would be something like:

`var results = quotes.Smooth(8).GetMfi(5)`

Smooth() in this case is a method that smooths out the erratic signals in this instance. I actually have a use case for this where I'd like to use a Mfi[13] and Mfi[34] as a mfi/signal crossover pair. however, as it stands currently, the trace is too crude to make really objective decisions from. A smoothing function would (in my mind) offer a better read (both visually and in code).

Just a thought, tho am of course aware that the Smooth() function would have to follow the GetMfi() call - dilemmas ...

**Reply by DaveSkender** [reply]
You're in luck.  I'm currently working on chaining indicators in #802 to make this easier.  "smooth" is usually just some sort of moving average.  For your example, you will be able to do something like:

```csharp
var results = quotes
  .GetMfi(..)
  .GetAlma(..);
```

**Reply by jimtollan** [reply]
well, i'll be darned!! zeitgeist or what :D

**Reply by jimtollan** [reply]
btw - OT. There are some nice references in the ThinkOrSwim lib. I quite liked this one tbh: https://tlc.thinkorswim.com/center/reference/Tech-Indicators/studies-library/C-D/CondensedCandles

https://sacredtraders.com/candlesticks-condensed-by-dave-cline/

**Reply by DaveSkender** [reply]
chaining is now available in v2.0.0

**Reply by jimtollan** [reply]
Dave - thanks for an amazing job on V2.0. Have only incorporated the chaining as per my query above, but can already see some big wins with other chained commands...

It just gets better and better...

**Reply by jimtollan** [reply]
Dave - apologies for spamming this thread with the following, but seemed like a reasonable place to push it for now. I know you have been looking at candle identification. i happened to stumble upon this: https://tickblaze.com/scripts/pattern

the code is pretty bad and there's lots of stupid things such as not using recursive methods (lots of multi level if's), plus no base candle type identified and reused. BUT - there's been quite a bit of thought applied as to what defines formations such as bullish-harami-cross, bullish-doji-star etc, etc. So might be a useful reference for your candles based stuff, even if just to discount the logic!!

**Reply by DaveSkender** [reply]
Thank you.  We're covering candlestick pattern discussions in #220
